073-improve-handling-of-CA-renewals-and-replacements-in-client-based-operands.md

```diff
@@ -0,0 +1,231 @@
+If you migrate to the new CA manually, you might still be able to orchestrate everything by editing the custom resource and modifying the trusted certificates as needed.
+
+But when the whole process is automated, it might not be easy to orchestrate it manually.
+This applies even in the ideal situation where the Apache Kafka cluster is based on Strimzi and runs in the same namespace.
+For example, the following happens when a user triggers CA replacement using the `strimzi.io/force-replace` annotation: 
+1. Strimzi generates the new CA public and private keys and stores the new CA public key as `ca.crt` in the Kubernetes Secret.
+   The public key of the old CA is renamed to `ca-<timestamp>.crt`.
+2. Strimzi starts rolling the Kafka brokers to make sure they trust the new CA.
+   At this point in time, the Kafka brokers still use the old server certificates signed by the old CA.
+3. The Kafka client based operand (e.g. Kafka Connect cluster) using the typical `.spec.tls` configuration:
+   ```yaml
+   tls:
+      trustedCertificates:
+         - secretName: my-cluster-cluster-ca-cert
+           certificate: ca.crt
+   ```
+   detects that the `ca.crt` in the Secret was updated and triggers a rolling update of the operand to use the new CA.
+   After the rolling update, it ceases to trust the old CA still in use by the Kafka brokers, resulting in the operand failing to connect to the brokers.
+4. After all brokers are rolled to trust the new CA, another rolling update happens and new server certificates signed by the new CA are rolled out to the Kafka brokers in another rolling update.
+5. Once all brokers are rolled once again and use the new CA, the operand such as Kafka Connect will be able to connect again and start working.
+   The impact and length of the outage will depend on many different aspects:
+   * How many nodes does the operand such as Kafka Connect have.
+     If it is only one replica, the whole operand will be down.
+     With multiple replicas, only one replica should be down.
+   * How big and busy the Kafka cluster is.
+     A small cluster might take few minutes to move to the new CA completely.
+     A big cluster with many nodes that are busy and need more time to re-sync between restarts might take much longer.
+6. After all brokers (and other components such as ZooKeeper, Topic and User Operator, Cruise Control etc.) use the new CA, another rolling update will be done to completely remove the trust to the old CA by removing the earlier `ca-<timestamp>.crt` from the Secret and rolling all components once again.
+   This last rolling update does not impact operands such as Kafka Connect.
+
+## Proposal
+
+This proposal suggests improving the API of our custom resources that we use for configuring the trusted certificates.
+It should allow loading more certificates without an exact specification of the keys in the Secret.
+To do so, we should add a new field called `certificatesMatching`.
```

Comment:
Key is not mentioned in the original `certificate` field either and it did not seem to be confusing to users.

Reply:
Who said it is confusing users? I just think it's a better name. Now you changed to pattern, but pattern of what? Key? Value? Both?

Reply:
> Who said it is confusing users?

So, what value does it add if it is not confusing and clear to everyone?

Reply:
Clarity? As I said, I think `keyPattern` is almost self describing. It could also be `keyGlobPattern`, but maybe this is too long.

Reply:
I thought some more about it, but to be honest, I do not see any reasons to use `keyPattern` instead of `pattern`.  I do not think it adds any value over `pattern`. So I decided to reject this, sorry.

Reply:
The value is better UX experience IMO. An alternative would be secretKeysRef(name, pattern), which would be even better.

Reply:
FTR I'm also -1 on using the word 'key'. The problem for me is that 'key' is an overloaded term. Someone might think it refers to the private key corresponding to the certificate. But it could mean the thing which identifies a pair within a `Secret`'s `data`. But if you use "secret key" that could be confused for a different concept from symmetric cryptography.

Reply:
In my last comment I proposed something like the following, which I think is better. I still think that just pattern is not great, but I can live with it.

```yaml
   tls:
      trustedCertificates:
         - secretKeysRef:
             name: my-cluster-cluster-ca-cert
             pattern: "*.crt"
```